### ISSUE TITLE
Dimension Blacklist

### ISSUE THREAD

**Issue by copygirl** [TelepathicGrunt/RepurposedStructures-Fabric#33]
It looks like some of Repurposed Structures' worldgen structures are generated in the [Heart of the Machine](https://github.com/Heart-of-the-Machine/heart-of-the-machine) dimension. Flying around I found jungle mineshafts and jungle fortresses. This could be due to how the "thinking forest" biome is tagged.

![2020-09-30_08 49 19](https://user-images.githubusercontent.com/272586/94652184-e2326d80-02f9-11eb-9e63-3f97fcd30f2b.png)

I didn't find an option to completely disable this mod's worldgen for specific dimensions in `repurposed_structures-main.toml`. Is it possible? If not, it would be lovely to have an option for that.

**Comment by TelepathicGrunt** [comment]
You would use the biome blacklist for those structures. A dimension blacklist really doesn’t work without some massive ugly hackery as structures are tied to biome and dimensions can share biomes.

Instead of editing the config files directly, put on mod menu and you can access the config gui. Then in each config section for each structure, put in all of Heart of the Machine’s biomes into the biome blacklist. When done, close and restart Minecraft and the change will take effect.

**Comment by copygirl** [comment]
In my case that is going to be sufficient, but I wonder what to do if the dimension had a ton of biomes (annoying to blacklist), or the dimension had Vanilla biomes.

**Comment by TelepathicGrunt** [comment]
Dimensions using vanilla biomes will still have the structures the biomes had in the overworld. This is the behavior that Mojang intended. Hence why I am hesitant to try and find a safe way to make structure not generate in the dimension. In fact, the actually class that generates the structure does not have access to the world and so, it can’t tell what dimension it’s in. 

An idea could be to mess with the chunk generator as there’s something in it that also hold structures but multiple dimensions can share chunk generators too. Basically Mojang seemed to try really hard to tie structures to biomes and partially to chunk generators. And also prevent structures from even being dimension based

**Comment by copygirl** [comment]
I see, that is perhaps a little unfortunate.

Additionally, something that would be nice is a global biome blacklist that applies to all features.

**Comment by TelepathicGrunt** [comment]
Version 1.6.2 Repurposed Structures is now out! The dimension blacklist is now an option in the Main category config menu. 

Thank you for bringing up the dimension blacklist as when I was working on a Structure Tutorial Mod for modders, I looked closer at the chunkgenerator as realized the behavior of it changed. Before, if a structure's spacing was not in it, it would spawn the structure in every single chunk. Now, if the chunk generator doesn't have the structure's spacing, it won't spawn it at all. I took advantage of this new change and made a hack to allow the dimension blacklist to work. :)

Well, in theory it should work lmao. I didn't test it a lot
